### PR TITLE
do not allow juju upgrade to 2.8 if controller is openstack using nova networking.

### PR DESCRIPTION
**Pull request by hmlanigan**
## Description of change

With juju 2.8.0, openstack clouds with nova networking are no longer supported.  Add a pre upgrade check to error if an upgrade to juju 2.8 is attempted on such a substrate.  This also adds a framework for other provider level prechecks of upgrades to be added.

## QA steps

We do not have a cloud using nova networking to tests against.  UnitTests in the openstack provider will have to be sufficient, for that case
```console
# bootstrap 2.6.10 microstack to ensure precheck upgrade step is not run for upgrade to 2.7.6.
# then upgrade to 2.8-rc1 to see the step run.
$ sudo snap refresh juju --channel 2.6/stable
$ /snap/bin/juju bootstrap --metadata-source <path>/simplestreams microstack micro-o7k-2.6

$ juju upgrade-controller --agent-version 2.7.6 --debug --show-log
# find "ignoring precheck upgrade operation for version 2.8.0" in output

$ juju upgrade-juju --debug --show-log --agent-version 2.7.6
# do not find "precheck upgrade operation" in output

$ juju upgrade-controller --build-agent --debug --show-log
# find 'running precheck step "Verify Neutron OpenStack service enabled" ' in output

$ juju upgrade-juju --debug --show-log --agent-version 2.8-rc1.1
# do not find "precheck upgrade operation" in output

# bootstrap lxd to ensure the recheck upgrade step is not run.
$ sudo snap refresh juju --channel 2.7/stable
$ /snap/bin/juju bootstrap --metadata-source <path>/simplestreams microstack micro-o7k

$ juju upgrade-controller --build-agent --debug --show-log
# do not find "running precheck upgrade operation for version 2.8.0" in output

$ juju upgrade-juju --debug --show-log --agent-version 2.8-rc1.1
# do not find "running precheck upgrade operation for version 2.8.0" in output

```

## Documentation changes

We need to note that openstack nova networking is not supported in juju 2.8.  This changes errors if the upgrade is attempted.

## Bug reference

https://bugs.launchpad.net/juju/+bug/1855345

### CODE REVIEW

File: provider/openstack/upgrades_test.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+import (
+	"github.com/golang/mock/gomock"
+	"github.com/juju/errors"
+	"github.com/juju/version"
+	gc "gopkg.in/check.v1"
+	"gopkg.in/goose.v2/identity"
+
+	"github.com/juju/juju/environs"
+	jc "github.com/juju/testing/checkers"
```
[the remark this answers is not in the document]
Comment (hmlanigan) [reply provider/openstack/upgrades_test.go]: fixed the import grouping

**Comment by hmlanigan** [comment]
@manadart, When failing an upgrade from the upgradedatabase worker, how easily are failures bubbled up to the user?

**Comment by hmlanigan** [comment]
!!build!!

**Comment by hmlanigan** [comment]
@manadart, another consideration I had was for upgrade model in a multi-cloud config where a single model is an openstack using nova networking.  This changes does not address that.  However I plan to see how to add it in.

**Comment by hmlanigan** [comment]
intermittent fail in WorkerSuite.TestErrRefresh

**Comment by hmlanigan** [comment]
!!build!!

**Comment by petevg** [comment]
I had some trouble w/ the QA steps, but I think that it's more missing knowledge on my part (and possibly https://bugs.launchpad.net/juju/+bug/1876366?), and not an issue w/ the code. Since it's running pretty late here, I'm going to go ahead and suggest that we merge it. I'll sync up w/ you tomorrow on what might have gone wrong w/ the QA run.

**Comment by petevg** [comment]
$$merge$$

**Comment by wallyworld** [comment]
$$merge$$

**Pull request merged**